domain/storage/service/storagepool.go

```diff
@@ -0,0 +1,246 @@
+	// Use the existing provider type unless explicitly overwritten.
+	if providerType == "" {
+		existingConfig, err := s.st.GetStoragePoolByName(ctx, name)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		providerType = storage.ProviderType(existingConfig.Provider)
+	}
+
+	err := s.validateConfig(name, providerType, attrs)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	attrsToSave := transform.Map(attrs, func(k string, v any) (string, string) { return k, fmt.Sprint(v) })
+	sp := domainstorage.StoragePoolDetails{
+		Name:     name,
+		Provider: string(providerType),
+		Attrs:    attrsToSave,
+	}
+	err = s.st.ReplaceStoragePool(ctx, sp)
+	return errors.Annotatef(err, "replacing storage pool %q", name)
+}
+
+// ListStoragePools returns the storage pools matching the specified filter.
+func (s *StoragePoolService) ListStoragePools(ctx context.Context, filter domainstorage.StoragePoolFilter) ([]*storage.Config, error) {
+	if err := s.validatePoolListFilter(filter); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	pools, err := domainstorage.BuiltInStoragePools()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	sp, err := s.st.ListStoragePools(ctx, filter)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	pools = append(pools, sp...)
+
+	results := make([]*storage.Config, len(pools))
+	for i, p := range pools {
+		results[i], err = s.toStorageConfig(p)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+	return results, nil
+}
+
+func (s *StoragePoolService) validatePoolListFilter(filter domainstorage.StoragePoolFilter) error {
+	if err := s.validateProviderCriteria(filter.Providers); err != nil {
+		return errors.Trace(err)
+	}
+	if err := s.validateNameCriteria(filter.Names); err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (s *StoragePoolService) validateNameCriteria(names []string) error {
+	for _, n := range names {
+		if !storage.IsValidPoolName(n) {
+			return fmt.Errorf("pool name %q not valid%w", n, errors.Hide(storageerrors.InvalidPoolNameError))
+		}
+	}
+	return nil
+}
+
+func (s *StoragePoolService) validateProviderCriteria(providers []string) error {
+	if s.registry == nil {
+		return errors.New("cannot filter storage providers without a registry")
+	}
+	for _, p := range providers {
+		_, err := s.registry.StorageProvider(storage.ProviderType(p))
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+// GetStoragePoolByName returns the storage pool with the specified name, returning an error
+// satisfying [storageerrors.PoolNotFoundError] if it doesn't exist.
+func (s *StoragePoolService) GetStoragePoolByName(ctx context.Context, name string) (*storage.Config, error) {
+	if !storage.IsValidPoolName(name) {
+		return nil, fmt.Errorf("pool name %q not valid%w", name, errors.Hide(storageerrors.InvalidPoolNameError))
```

Comment:
I see you've turned to the dark side:smile:  I wonder if we should have `errors.Errorf` do this AND we get the benefit of `errors.Trace`?